lib/dockerregistry/checks_test.go

```diff
@@ -317,3 +318,113 @@ func TestImageSizeCheck(t *testing.T) {
 			fmt.Sprintf("checkError: test: %v (ImageSizeCheck)\n", test.name))
 	}
 }
+
+// TestImageVulnCheck uses a fake populateRequests function and a fake
+// vulnerability producer. The fake vulnerability producer simply returns the
+// vulnerability occurrences that have been mapped to a given PromotionEdge in
+// order to simulate running the real check without having to make an api call
+// to the Container Analysis Service.
+func TestImageVulnCheck(t *testing.T) {
+	edge1 := reg.PromotionEdge{
+		Digest: "foo",
+	}
+	edge2 := reg.PromotionEdge{
+		Digest: "bar",
+	}
+
+	mkVulnProducerFake := func(
+		edgeVulnerabilities map[reg.Digest][]*grafeaspb.VulnerabilityOccurrence,
+	) reg.ImageVulnProducer {
+		return func(
+			edge reg.PromotionEdge,
+		) ([]*grafeaspb.VulnerabilityOccurrence, error) {
+			return edgeVulnerabilities[edge.Digest], nil
+		}
+	}
+
+	var tests = []struct {
+		name              string
+		severityThreshold int
+		edges             map[reg.PromotionEdge]interface{}
+		vulnerabilities   map[reg.Digest][]*grafeaspb.VulnerabilityOccurrence
+		expected          error
+	}{
+		{
+			"Severity under threshold",
+			int(grafeaspb.Severity_HIGH),
+			map[reg.PromotionEdge]interface{}{
+				edge1: nil,
+				edge2: nil,
+			},
+			map[reg.Digest][]*grafeaspb.VulnerabilityOccurrence{
+				"foo": {
+					{
+						Severity: grafeaspb.Severity_LOW,
+					},
+				},
+				"bar": {
+					{
+						Severity: grafeaspb.Severity_MEDIUM,
+					},
+				},
+			},
+			nil,
+		},
+		{
+			"Severity at threshold",
+			int(grafeaspb.Severity_HIGH),
+			map[reg.PromotionEdge]interface{}{
+				edge1: nil,
+				edge2: nil,
+			},
+			map[reg.Digest][]*grafeaspb.VulnerabilityOccurrence{
+				"foo": {
+					{
+						Severity: grafeaspb.Severity_HIGH,
+					},
+				},
+				"bar": {
+					{
+						Severity: grafeaspb.Severity_HIGH,
+					},
+				},
+			},
+			fmt.Errorf("Encountered an error while executing requests"),
```

Comment:
Is there any way to make these errors less vague?

I see that this error message is coming from ExecRequests... I think we can improve this without touching ExecRequests.

Reply:
Done

Reply:
Improving this error message actually makes it easier to test some edge cases. For example, making sure that if two promotion edges have the same src image, only one of those edges is actually tested for vulnerabilities